test/jdk/javax/management/Introspector/ClassLeakTest.java

```diff
@@ -48,38 +49,10 @@ public static void main(String[] args) throws Exception {
                            "Standard MBean does not retain a reference to " +
                            "the MBean's class");
 
-
-        String[] cpaths = System.getProperty("test.classes", ".")
-                                .split(File.pathSeparator);
-        URL[] urls = new URL[cpaths.length];
-        for (int i=0; i < cpaths.length; i++) {
-            urls[i] = Paths.get(cpaths[i]).toUri().toURL();
-        }
-
-        @SuppressWarnings("removal")
-        PrivateMLet mlet = new PrivateMLet(urls, null, false);
-        Class<?> shadowClass = mlet.loadClass(TestMBean.class.getName());
-        if (shadowClass == TestMBean.class) {
-            System.out.println("TEST INVALID: MLet got original " +
-                               "TestMBean not shadow");
-            System.exit(1);
-        }
-        shadowClass = null;
-
         MBeanServer mbs = MBeanServerFactory.createMBeanServer();
-        ObjectName mletName = new ObjectName("x:type=mlet");
-        mbs.registerMBean(mlet, mletName);
-
-        ObjectName testName = new ObjectName("x:type=test");
-        mbs.createMBean(Test.class.getName(), testName, mletName);
-
-        ClassLoader testLoader = mbs.getClassLoaderFor(testName);
-        if (testLoader != mlet) {
-            System.out.println("TEST INVALID: MBean's class loader is not " +
-                               "MLet: " + testLoader);
-            System.exit(1);
-        }
-        testLoader = null;
+        ObjectName testName = new ObjectName("x:name=Test");
+        Test mbean = new Test();
+        mbs.registerMBean(mbean, testName);
```

Comment:
I didn't see great value in it, as we no longer provide an MBean which is a ClassLoader.
Having a test MBean that extends URLClassLoader, and calling its loadClass()... is basically testing URLClassLoader? 8-)
Implementing PrivateClassLoader affects ClassLoaderRepository behaviour, but that's not tested here.
If you think this have value, we can do it.

Reply:
My thinking is that this tests for a leak when an MBean is created using a ClassLoader which is also an MBean. MLet was such a ready-to-use ClassLoader. We no longer have MLets, but it's still possible to register an MBean that is a class loader and use that to create another MBean whose concrete class gets loaded by that ClassLoader MBean. We're testing that the Introspector doesn't create a leak in this case.

Reply:
Checking on the original problem noted in the test, looks like it wasn't really about ClassLoaders and MLets:
JDK-4909536: Standard MBean introspector keeps reference to last class introspected
This is what the test still tests.
Do we see value in having the MBean be a ClassLoader and checking if such MBeans are held on to...? 8-)

Reply:
Yes, that's the whole point of the test. Without using the an MBean which is a ClassLoader the fix cannot be tested.
1. you need to create an MBean which is an URLClassLoader, and it needs to be a PrivateClassLoader so that it's not added to the ClassLoaderRepository - let's call it  LoaderMBean.
2. you need to use that LoaderMBean to load another MBean - let's call it TestMBean. The concrete class of TestMBean needs to be loaded by the LoaderMBean - that is - LoaderMBean needs to be the defining ClassLoader for that class (the TestMBean class must not be loaded by the System/Application ClassLoader).
3. In the MBeanServer you then end up with two MBeans, one is the LoaderMBean, the other is the TestMBean whose class has been loaded by the LoaderMBean.
4. you then keep a weak reference on the _LoaderMBean_, and unregister both MBeans.
5. at that point - if the Introspector still has a strong reference to the class of the TestMBean, then the _LoaderMBean WeakReference_ will never be enqueued, because the class of the TestMBean will have a strong reference to its ClassLoader, which is _LoaderMBean_.

So you can only test that if you have an MBean which is loaded by a custom ClassLoader, because you want to verify that the Introspector doesn't hold on that ClassLoader (through the TestMBean class).

Reply:
Hi Daniel -
I am missing see how being a ClassLoader relates to the leak where the Introspector holds on to a reference to the most recent MBean.  I didn't locate the jdk5 change, I only read a short synopsis.

I can make this test create an MBean which is a ClassLoader, and the test can do its check that having the loader MBean load a class, creates a distinct class.  I can add that update herre.  But it looks like an MLet test which leaked into the Introspector test. 8-)

Reply:
To verify whether the Introspector holds a reference to a *Class Object* you need that class to be loaded by a custom class loader, and verify that the custom class loader that loaded that class has been garbage collected.
Hi Kevin,

If the class loader that loaded that class has not been garbage collected after you released all references to it, then it's because the class it loaded is still referenced somewhere.

The logic of the test is as follows:
You create an MBean L which is a class loader and able to load class C. L must be the defining class loader for C.
You register L in the MBeanServer.
You ask the MBeanServer to create an MBean of class C through L (this is why you must use the variant of createMBean that takes a loader name).
You keep a weak ref to L
You unregister L and the MBean of class C from the MBeanServer and make sure the test doesn't have any strong reference to them (or to class C).
Then you start GC and wait until the weak ref to L is enqueued. If that doesn't happen, it's because the Introspector still hold a reference to class C.

You cannot test that if C has been loaded by the system class loader.

Reply:
I am updating those two tests.
I read the test as checking that an Object instance is collected, rather than Class Object, so was favouring making it simpler.
But we can use an MBean which is a URLClassLoader, and things will be more similar to the originals.

Some of MXBeanLoadingTest1 is duplicating some of ClassLeakTest, so something for future auditing perhaps...